### ISSUE TITLE
v0.2 design won't work

### ISSUE THREAD

**Issue by pcpthm** [llogiq/compact_arena#3]
v0.2 design won't work as following example compiles:
```rust
    let escape = |idx| in_arena!(arena, {
        if let Some(idx) = idx {
            println!("{}", arena[idx]);
            idx
        } else {
            arena.add("hello")
        }
    });
    escape(Some(escape(None)));
```
We want to prevent escaping tag type from inner function of `in_arena!` but there is no way to express that in current Rust as far as I know.
Furthermore, escaping cannot prevent by the type system alone (not using lifetimes):
```rust
    fn recursive(idx: Option<Box<dyn std::any::Any>>) {
        in_arena!(arena, {
            if let Some(idx) = idx {
                println!("{}", arena[*idx.downcast().unwrap()]);
            } else {
                recursive(Some(Box::new(arena.add("hello"))));
            }
        });
    }
    recursive(None);
```

**Comment by llogiq** [comment]
Cool, you found a counter example! Thank you so much!

I will probably remove the macro and leave the unsafe functions in place. That and possibly add a bounds-checked safe variant.

**Comment by llogiq** [comment]
Thinking a bit more about the issue, I may have an idea how to preclude this at runtime: Keep a static AtomicBool with each Tag type, fetch & set to true on instantiation. If the fetched bool is true, panic.

**Comment by llogiq** [comment]
Thinking a bit more, we also need to set our bool to false when we drop the Arena. Otherwise our code could only run once.

**Comment by llogiq** [comment]
A threadlocal bool would be better than an AtomicBool, too.

**Comment by pcpthm** [comment]
If I'm not mistaken, adding mutex guard doesn't prevent us from running my first example.
The arena is already dropped when first `escape(None)` is returned.

**Comment by llogiq** [comment]
So we have to ensure no index escapes the closure.

**Comment by pcpthm** [comment]
Yes, and that condition can be represented with hypothetical hyper-ranked trait bounds syntax:
```rust
fn in_arena<R, C>(closure: C) -> R where C: for<B> FnOnce(Arena<B>) -> R;
```
Haskell's `ST` monad uses such signature. However, the above is
1. Not available in Rust.
2. Still vulnerable to type erasure `Any` as shown in my second example.

What we can do in Rust's type system is using lifetimes:
```rust
fn in_arena<R, C>(closure: C) -> R where C: for<'b> FnOnce(Arena<'b>) -> R;
```
This works because:
1. Higher ranked bound is supported for lifetimes.
2. `Any` requires `'static` bound and thus lifetimes cannot be escaped.

**Comment by pcpthm** [comment]
Maybe this work?
```rust
trait WithArena {
   type Output: Sized;
   fn with_arena<B>(arena: Arena<B>) -> Self::Output;
}
```

**Comment by llogiq** [comment]
I have found a way to bind the indices to a tag lifetime without using a closure (which to me seems beneficial as it allows using `return`s and `?` within the in_arena expression.

I just published 0.2.1 with the changes. I'm not entirely sure if it's sound now, but at least the indices can no longer escape.

**Comment by pcpthm** [comment]
Great! I think v0.2.1 is sound. Closing as the problem is now solved.
